### PR TITLE
RSDK-6883 rpc: Replace RS256 with EdDSA as default signing algo

### PR DESCRIPTION
**Pull request by edaniels**
- And support multiple keys for verification and select one for signing

### CODE REVIEW

File: rpc/auth.go [hunk 1]
```diff
@@ -256,18 +270,6 @@ type Credentials struct {
 	Payload string          `json:"payload"`
 }
 
-// RSAPublicKeyThumbprint returns SHA1 of the public key's modulus Base64 URL encoded without padding.
-func RSAPublicKeyThumbprint(key *rsa.PublicKey) (string, error) {
```
Comment (maximpertsov) [inline rpc/auth.go:260]: we should deprecate instead of removing - app is using this method.

**Review by maximpertsov — COMMENTED** [review]
looks good generally - we need to update the two examples (echo + fileupload) to use `ed25519`

File: rpc/auth.go (hunk 1 above)
Comment (edaniels) [reply rpc/auth.go:260]: it's using it via goutils/rpc, not app

Comment (edaniels) [reply rpc/auth.go:260]: ah right

**Comment by edaniels** [comment]
@maximpertsov ready again!

**Review by maximpertsov — APPROVED** [review]
LGTM as far as how this affects the RDK server. [Added a reviewer from fleet](https://github.com/viamrobotics/goutils/pull/257#issuecomment-1995052863) to make sure there are no concerns app-side.

**Comment by michaellee1019** [comment]
Sorry for the delay. Makes sense. Will we be running both RSA + EdDSA on viam's authentication backend for some time? I think we'll need to respect the already issued JWTs for some time. And then the plan is to switch to RDK to use EdDSA in the next release?

**Comment by edaniels** [comment]
@michaellee1019 what this PR does is sign new signatures in EdDSA but continues to accept all valid signature types. That means pre-deploy you have RSA signed JWTs. Post-deploy you have a mix. Weeks later without any deploy due to JWT expiration, you will have all EdDSA

**Comment by edaniels** [comment]
> @michaellee1019 what this PR does is sign new signatures in EdDSA but continues to accept all valid signature types. That means pre-deploy you have RSA signed JWTs. Post-deploy you have a mix. Weeks later without any deploy due to JWT expiration, you will have all EdDSA

Actually for app I think we can stick to RSA right now. I need to add back in support here. I was only considering RDK signed JWTs at the moment. I'll get to this this week.

**Comment by michaellee1019** [comment]
> Actually for app I think we can stick to RSA right now. I need to add back in support here. I was only considering RDK signed JWTs at the moment. I'll get to this this week.

Ok cool. TY!

**Comment by edaniels** [comment]
@maximpertsov @michaellee1019 need a re-review. I've now added support for multiple keys to be added and let the user select a key to use for signing if more than one has been provided. I realized it was silly to just support either/or and this wasn't much more code to add. This will allow the app server to utilize multiple verification keys and select a new signing key/method in the future which is good for security + flexibility!

Will add tests later this week and verify rdk/app acceptance.

**Comment by edaniels** [comment]
Tested this works when upgrading RDK and App (requires a one line change due to breaking API). RDK doesn't really matter much since every server restart is already a new private key. App will continue to work as is but can be extended to support both types of keys in the future for a no-downtime swap to EdDSA (if desired).

**Pull request merged**